src/Agent.Listener/Configuration/FeatureFlagProvider.cs

```diff
@@ -0,0 +1,117 @@
+using Microsoft.VisualStudio.Services.Agent.Util;
+using System;
+using System.Net.Http;
+using System.Net.Http.Json;
+using System.Runtime.Serialization;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using System.Threading.Tasks;
+
+namespace Agent.Listener.Configuration
+{
+    [ServiceLocator(Default = typeof(FeatureFlagProvider))]
+    public interface IFeatureFlagProvider : IAgentService
+    {
+        /// <summary>
+        /// Gets the status of a feature flag from the specified service endpoint.
+        /// If request fails, the feature flag is assumed to be off.
+        /// </summary>
+        /// <param name="serviceEndpoint">The service endpoint to get the feature flag status from.</param>
+        /// <param name="featureFlagName">The name of the feature flag to get the status of.</param>
+        /// <returns>The status of the feature flag.</returns>
+        public Task<FeatureFlagStatus> GetFeatureFlagAsync(IHostContext hostContext, ServiceEndpoint serviceEndpoint, string FeatureFlagName, ITraceWriter traceWriter);
+    }
+
+    public class FeatureFlagProvider : AgentService, IFeatureFlagProvider, IDisposable
+    {
+        private HttpClient _httpClient;
+
+        public FeatureFlagProvider()
+        {
+            
+        }
+
+        public FeatureFlagProvider(HttpClient client) 
+        {
+            _httpClient = client;
+        }
+
+        public async Task<FeatureFlagStatus> GetFeatureFlagAsync(IHostContext hostContext, ServiceEndpoint serviceEndpoint, string featureFlagName, ITraceWriter traceWriter)
+        {
+            ArgUtil.NotNull(serviceEndpoint, nameof(serviceEndpoint));
+            ArgUtil.NotNull(featureFlagName, nameof(featureFlagName));
+            using HttpClientHandler httpHandler = hostContext.CreateHttpClientHandler(); ;
+            if (_httpClient == null)
+            {
+                _httpClient = new HttpClient(httpHandler);
+            }
+            _httpClient.Timeout = TimeSpan.FromSeconds(10);
+            var accessTokenFF = serviceEndpoint?.Authorization?.Parameters["AccessToken"] ?? null;
+            var url = $"{serviceEndpoint.Url}_apis/FeatureFlags/{featureFlagName}";
+            _httpClient.DefaultRequestHeaders.Add("Authorization", $"Bearer {accessTokenFF}");
+            var response = await _httpClient.GetAsync(url);
+            if (response.IsSuccessStatusCode)
+            {
+                try
+                {
+                    return await response.Content.ReadFromJsonAsync<FeatureFlagStatus>();
+                } catch(JsonException ex)
+                {
+                    traceWriter.Verbose($"Failed to parse JSON with exception: {ex.Message}");
+                }
+            }
+            else
+            {
+                traceWriter.Verbose($"Failed to retrive FeatureFlag {featureFlagName} with error code: {response.StatusCode}");
+            }
+
+            return new FeatureFlagStatus()
```

Comment:
fallback on timeout should not be to return the wrong values- that could lead to the wrong code being executed